homeassistant/helpers/entity.py

```diff
@@ -315,9 +317,52 @@ def unique_id(self) -> str | None:
         """Return a unique ID."""
         return self._attr_unique_id
 
+    @callback
+    def async_full_name(self, device_entry: dr.DeviceEntry) -> str | None:
+        """Combine device name with entity name."""
+
+        def _compile_name(entity_name: str | None) -> str | None:
+            """Help combine device name with entity name."""
+            if not entity_name:
+                return device_entry.name_by_user or device_entry.name
+            return f"{device_entry.name_by_user or device_entry.name} {entity_name}"
+
+        if hasattr(self, "_attr_entity_name"):
+            return _compile_name(self._attr_entity_name)
+        if (
+            hasattr(self, "entity_description")
+            and (entity_name := self.entity_description.entity_name) is not UNDEFINED
+        ):
+            return _compile_name(entity_name)
+        if (entity_name := self.entity_name) is not UNDEFINED:
+            return _compile_name(entity_name)
+        return None
+
+    def _full_name(self) -> str | None:
+        """Combine device name with entity name."""
+
+        if not self.registry_entry:
+            return None
+
+        device_registry = dr.async_get(self.hass)
+        if not (device_id := self.registry_entry.device_id) or not (
+            device := device_registry.async_get(device_id)
+        ):
+            return None
+
+        return self.async_full_name(device)
+
+    @property
+    def entity_name(self) -> str | None | UndefinedType:
+        """Return the name of the entity."""
+        return UNDEFINED
```

Comment:
When do we want to implement this property and return None explicitly? Let's add that explanation to the docstring.